torch/csrc/jit/codegen/cuda/shape_expr_memo.h

```diff
@@ -0,0 +1,152 @@
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+
+#include "glfdc/eval.h"
+#include "glfdc/expr_builder.h"
+
+#include <unordered_map>
+
+namespace glfdc {
+struct ExpressionBuilder;
+struct EvalState;
+} // namespace glfdc
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace kir {
+
+class ExprBuilderVisitor : IrVisitor {
+ public:
+  explicit ExprBuilderVisitor(
+      glfdc::ExpressionBuilder& bld,
+      const Kernel& kernel);
+
+  c10::optional<glfdc::Expr> build(const Val* value) {
+    auto ref = build_operand(value);
+    return builder_->create_expr(ref);
+  }
+
+  glfdc::Operand build_operand(const Val* value) {
+    TORCH_CHECK(value->isScalar());
+    TORCH_CHECK(value->dtype() == DataType::Int);
+
+    if (value->isA<Int>()) {
+      if (value->isConst()) {
+        auto opt_val = value->as<Int>()->value();
+        if (opt_val.has_value())
+          return glfdc::Value(opt_val.value());
+        else
+          return builder_->get_binding(reinterpret_cast<uintptr_t>(value));
+      }
+    }
+
+    current_operand_ = c10::nullopt;
+    value->accept(this);
+    TORCH_CHECK(current_operand_.has_value());
+    auto ret = current_operand_.value();
+    current_operand_ = c10::nullopt;
+    return ret;
+  }
+
+  auto takeDAG() noexcept {
+    return builder_->takeDAG();
+  }
+
+  auto reuses() noexcept {
+    return builder_->reuses();
+  }
+
+ private:
+  void unhandled(const void*) final;
+  void visit(const Int* value) final;
+  void visit(const NamedScalar* named_scalar) final;
+  void visit(const UnaryOp* unary_op) final;
+  void visit(const BinaryOp* binary_op) final;
+
+  glfdc::ExpressionBuilder* builder_;
+  c10::optional<glfdc::Operand> current_operand_ = c10::nullopt;
+  const Kernel* kernel_;
+};
+
+inline ExprBuilderVisitor::ExprBuilderVisitor(
+    glfdc::ExpressionBuilder& bld,
+    const Kernel& kernel)
+    : builder_(&bld), kernel_(&kernel) {}
+
+inline void ExprBuilderVisitor::unhandled(const void*) {
+  TORCH_INTERNAL_ASSERT(
+      false, "Kernel IR expression evaluation reached an unsupported node");
+}
+
+inline void ExprBuilderVisitor::visit(const Int* value) {
+  TORCH_INTERNAL_ASSERT(!value->isConst());
+  if (auto def = value->definition()) {
+    def->accept(this);
+  } else if (kernel_->isInput(value)) {
+    TORCH_CHECK(false && "Unhandled input scalar");
+  }
+}
+
+inline void ExprBuilderVisitor::visit(const UnaryOp* unary_op) {
+  glfdc::Operand oper = build_operand(unary_op->in());
+  switch (unary_op->operation()) {
+    case UnaryOpType::Neg:
+      current_operand_ = builder_->create_sexpr(
+          glfdc::OperatorKind::sub, glfdc::Value(0), oper);
+      break;
+    case UnaryOpType::Cast:
+      current_operand_ = oper;
+    default:
+      break;
+  }
+
+  TORCH_INTERNAL_ASSERT(
+      current_operand_.has_value() && "Unexpected operator type");
+}
+
+inline void ExprBuilderVisitor::visit(const BinaryOp* binary_op) {
```

Comment:
I'd prefer functions like this placed in a source file. The possible gain by inlining would be negligible compared to the time to execute the function body.

Reply:
Sure

Reply:
All moved to cpp file

Reply:
I still see some heavy header files in the glfdc directory.